### PR TITLE
    [ceph] Add more data collection to ceph_osd

### PR DESCRIPTION
**Pull request by nkshirsagar**
Enhance the ceph_osd plugin to collect more data
    specific to OSD nodes.

    Closes: #2735

    Signed-off-by: Nikhil Kshirsagar <nkshirsagar@gmail.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

### CODE REVIEW

File: sos/report/plugins/ceph_osd.py [hunk 1]
```diff
@@ -39,9 +40,47 @@ def setup(self):
 
         self.add_cmd_output([
             "ceph-disk list",
-            "ceph-volume lvm list",
         ])
 
+        ceph_cmds = [
```
Comment (hillpd) [inline sos/report/plugins/ceph_osd.py:45]: This list looks good. Please add the following two commands:
"version",
"perf histogram dump",

For details, see: https://docs.ceph.com/en/latest/dev/perf_histograms/

File: sos/report/plugins/ceph_osd.py [hunk 2]
```diff
@@ -39,9 +40,47 @@ def setup(self):
...
+           "dump_watchers",
+           "log dump",
+           "perf dump",
+           "objecter_requests",
+           "ops",
+           "status"
+        ]
+
+        osd_ids = []
+        output = self.collect_cmd_output('ceph-volume lvm list --format json')
```
Comment (hillpd) [inline sos/report/plugins/ceph_osd.py:67]: There have been a number of performance issues with ceph-volume in the past. I do not recommend using this command to build a list of osd ids. 

It is a bit uglier, but you can build a list of ids by parsing `mount` output. As an example, this is the command systemd uses to generate a list of the ceph-osd services that need to be started:
`sudo mount|grep /var/lib/ceph/osd/ceph-|awk "{print \$3}"|cut -d- -f2|sort -n|xargs`

**Review by hillpd — CHANGES_REQUESTED** [review]

File: sos/report/plugins/ceph_osd.py (hunk 2 above)
Comment (pponnuvel) [reply sos/report/plugins/ceph_osd.py:67]: An equally ugly but slightly more efficient version of it can be
```
sudo mount|awk -v p=/var/lib/ceph/osd/ceph- '$0~p{split($3, x, "-"); printf x[2]" "}'
```
(sorting isn't necessary here).

Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:67]: Hi @pponnuvel  , @hillpd 

thank you for the review and suggestions!

The collect_cmd_output() method provided by sosreport does not allow pipes. So we need to get the output of a command, then implement the output extraction in the sos plugin.

Another approach might be to extract it from output of ps -u ceph -o args, and extract the id numbers that way (The ceph-osd process belong to the user ceph). 

output of "ps -u ceph -o args" looks like - 

COMMAND
/usr/bin/ceph-osd -f --cluster ceph --id 0 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 1 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 2 --setuser ceph --setgroup ceph

If we use this approach, we would not collect data for an OSD that is stopped by admin, or even down/crashed, possibly saving the plugin some time. Please let me know what you think.

Comment (TurboTurtle) [reply sos/report/plugins/ceph_osd.py:67]: Chiming in: pipes and bash foo in general aren't allowed in plugins.

You could do something like the following, if `ceph-volume lvm list` is not desired:

```python
out = self.collect_cmd_output('mount')
mounts = [
    m.split()[2] for m in out['output'].splitlines() if 'ceph/osd/ceph-' in m
]
```

However, the actual collection of `ceph-volume lvm list` seems important as it's been in sos for a while. What kind of performance concerns are there? Is it just related to the `ceph-volume` command, or does this affect the entire host? If we're going to keep collecting the output of `ceph-volume lvm list`, I have no issues with using it to build the list.

We have several safe guards in place to prevent any under-performing command collection from impacting the rest of the plugin, or the report as a whole, too much. At worst, we'd hit a timeout and report that timeout to the user, then continue on as normal.

File: sos/report/plugins/ceph_osd.py (hunk 1 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:45]: ack done

File: sos/report/plugins/ceph_osd.py [hunk 3]
```diff
@@ -39,9 +40,47 @@ def setup(self):
...
+        containers_list = self.get_all_containers_by_regex("ceph-osd*")
+
+        if containers_list:
+            self.add_cmd_output([self.fmt_container_cmd(
+                containers_list[0][1], "ceph daemon %s %s"
+                % (osid, cmd)) for osid in osd_ids for cmd in ceph_cmds])
+
+        else:
+            self.add_cmd_output(["ceph daemon %s %s" % (
+                osid, cmd) for osid in osd_ids for cmd in ceph_cmds])
```
[the remark this answers is not in the document]
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:82]: ack, thanks.

File: sos/report/plugins/ceph_osd.py (hunk 2 above)
Comment (hillpd) [reply sos/report/plugins/ceph_osd.py:67]: @nkshirsagar 
I think parsing either `ps` or `mount` output is a reasonable approach.

@TurboTurtle
I'm not suggesting that we stop collecting `ceph-volume lvm list`. We should just avoid using this command to build the `osd_ids` list. A time-out would prevent all of the `ceph_cmds` from being collected on lines 78 and 82.

> What kind of performance concerns are there? Is it just related to the `ceph-volume` command, or does this affect the entire host?

My concern is that the `ceph-volume` command may be prone to time-out on hosts running older Ceph code levels that sosreport can still target. There are several bugs I'm concerned about: [lp#1850754](https://bugs.launchpad.net/ubuntu/+source/ceph/+bug/1850754) and [lp#1908375](https://bugs.launchpad.net/ubuntu/+source/ceph/+bug/1908375).

**Review by slashdd — APPROVED** [review]
lgtm

File: sos/report/plugins/ceph_osd.py [hunk 4]
```diff
@@ -39,9 +39,62 @@ def setup(self):
...
+           "ops",
+           "status",
+           "version",
+        ]
+
+        osd_ids = []
+        out = self.exec_cmd('ps -u ceph -o args')
+
+        if out['status'] != 0:
+            return
```
Comment (pmoravec) [inline sos/report/plugins/ceph_osd.py:72]: This will mean **whole remaining code in `setup()`** will be ignored and no command output from below commands will be collected - is that intentional?

File: sos/report/plugins/ceph_osd.py [hunk 5]
```diff
@@ -39,9 +39,62 @@ def setup(self):
...
+
+        # Get the running OSD process IDs from ps
+        osd_nums = [
+                m.split()[5] for m in out['output'].splitlines()
+                if 'ceph --id' in m
+        ]
+
+        for element in osd_nums:
+            osd_ids.append('osd.' + element)
```
Comment (TurboTurtle) [inline sos/report/plugins/ceph_osd.py:81]: Shorten this up into a single list:

```python
if out['status'] == 0:
    osd_ids = [
        'osd.%s' % m.split()[5] for m in out['output'].splitlines()
        if 'ceph --id' in m
    ]
else:
    osd_ids = []
```

File: sos/report/plugins/ceph_osd.py [hunk 6]
```diff
@@ -39,9 +39,62 @@ def setup(self):
...
+
+        osd_ids = []
+        out = self.exec_cmd('ps -u ceph -o args')
+
+        if out['status'] != 0:
+            return
+
+        # Get the running OSD process IDs from ps
+        osd_nums = [
+                m.split()[5] for m in out['output'].splitlines()
```
Comment (pmoravec) [inline sos/report/plugins/ceph_osd.py:76]: Are we sure each line will have at least 6 words, and no `IndexError` exception can be raised?

File: sos/report/plugins/ceph_osd.py [hunk 7]
```diff
@@ -39,9 +39,62 @@ def setup(self):
...
+
+        for element in osd_nums:
+            osd_ids.append('osd.' + element)
+
+        containers_list = self.get_all_containers_by_regex("ceph-osd*")
+
+        if containers_list:
+            self.add_cmd_output([
+                self.fmt_container_cmd(
+                    containers_list[1][1], "ceph daemon %s %s"
```
Comment (pmoravec) [inline sos/report/plugins/ceph_osd.py:88]: Can't `containers_list` contain exactly one element / one container? (if so, also `IndexError` could happen)

File: sos/report/plugins/ceph_osd.py (hunk 4 above)
Comment (TurboTurtle) [reply sos/report/plugins/ceph_osd.py:72]: Agreed, we don't need to return here. If the command status is non-zero, then the following logic to populate osd_ids can be skipped, which will in turn make the `add_cmd_output()` calls a no-op. But, by not issuing a return here, we still hit the `add_forbidden_path()` call, without needing to move it around as well within the flow of `setup()`.

File: sos/report/plugins/ceph_osd.py (hunk 7 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:88]: agreed. this crept in through the changes made for https://github.com/sosreport/sos/pull/2736#discussion_r738476601 .. I guess 

self.fmt_container_cmd(
                    containers_list[0][1], "ceph daemon %s %s"

would work, since we just need any one container so the first one is as good as any.

File: sos/report/plugins/ceph_osd.py (hunk 5 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:81]: ack thank you

File: sos/report/plugins/ceph_osd.py (hunk 6 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:76]: Yes I think we can expect this output to be consistent like the format I see today, but I will check again with @pponnuvel or @hillpd if this might not be the case. I'd expect the id number always to be element number 6.

root@bronzor:~/sos/sos# ps -u ceph -o args
COMMAND
/usr/bin/ceph-osd -f --cluster ceph --id 0 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 1 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 2 --setuser ceph --setgroup ceph

File: sos/report/plugins/ceph_osd.py (hunk 4 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:72]: done, thanks.

File: sos/report/plugins/ceph_osd.py (hunk 6 above)
Comment (pponnuvel) [reply sos/report/plugins/ceph_osd.py]: It may be possible that there are other ceph processes running on the same host, not just OSDs. So it's possible to hit `IndexError`.

This could be handled by checking:
- the line contains at least 6 fields
- 5th field is `--id`
- 6th field is a number (not a string e.g. mon which could be running on the same machine in a hyperconverged env)

Which would rule out others except OSDs.

Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py]: ack @pponnuvel , done

File: sos/report/plugins/ceph_osd.py [hunk 8]
```diff
@@ -39,9 +39,60 @@ def setup(self):
...
+        osd_ids = []
+        out = self.exec_cmd('ps -u ceph -o args')
+
+        if out['status'] == 0:
+            # Get the running OSD process IDs from ps
+            osd_ids = [
+                'osd.%s' % m.split()[5] for m in out['output'].splitlines()
+                if 'ceph --id' in m and len(m.split()) > 5
+                and m.split()[5].isdigit() and m.split()[4] == "--id"
+            ]
```
Comment (pmoravec) [inline sos/report/plugins/ceph_osd.py:77]: We are redundantly calling `m.split()` many times in one iteration of `for m in ..`. But calling it once would mean less obvious/elegant code and I hope/expect the data are kept in some python's/kernel cache, and the "`ps`" output isnt too big.

TL;DR: I am not happy with this, but I see it as best option.

File: sos/report/plugins/ceph_osd.py [hunk 9]
```diff
@@ -39,9 +39,60 @@ def setup(self):
...
+        else:
+            osd_ids = []
```
Comment (pmoravec) [inline sos/report/plugins/ceph_osd.py:79]: Dead code due to "`osd_ids = []`" in line 68 :)

**Review by pmoravec — APPROVED** [review]
ACK even for the current code (removing the "dead"(*) code is preferred, though).

(*) it is rather no-op than dead, program execution *can* get there but will apply nothing.

File: sos/report/plugins/ceph_osd.py (hunk 8 above)
Comment (TurboTurtle) [reply sos/report/plugins/ceph_osd.py:77]: If we're going to expand the safeguard checks this much, we don't _need_ to keep everything in a single list constructor. The request to do this originally was because there was only a single check occurring and there were 2 lists being used to create the single list of osd ids.

```python
osd_ids = []
for procs in out['output'].splitlines():
    proc = procs.split()
    if len(proc) < 6:
        continue
    if proc[4] == '--id' and proc[5].isdigit():
        osd_ids.append("osd.%s" % proc[5])
```

Single `split()` call, and is very easily grok'd.

Comment (pmoravec) [reply sos/report/plugins/ceph_osd.py:77]: +1 to this.

File: sos/report/plugins/ceph_osd.py (hunk 9 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:79]: removed, thanks.

File: sos/report/plugins/ceph_osd.py (hunk 8 above)
Comment (nkshirsagar) [reply sos/report/plugins/ceph_osd.py:77]: leaving as is, thank you

**Review by pmoravec — APPROVED** [review]
> If we're going to expand the safeguard checks this much, we don't _need_ to keep everything in a single list constructor. The request to do this originally was because there was only a single check occurring and there were 2 lists being used to create the single list of osd ids.
> 
> ```python
> osd_ids = []
> for procs in out['output'].splitlines():
>     proc = procs.split()
>     if len(proc) < 6:
>         continue
>     if proc[4] == '--id' and proc[5].isdigit():
>         osd_ids.append("osd.%s" % proc[5])
> ```
> 
> Single `split()` call, and is very easily grok'd.

I would prefer to have this applied, but ACK even the current code.

**Comment by nkshirsagar** [comment]
> > If we're going to expand the safeguard checks this much, we don't _need_ to keep everything in a single list constructor. The request to do this originally was because there was only a single check occurring and there were 2 lists being used to create the single list of osd ids.
> > ```python
> > osd_ids = []
> > for procs in out['output'].splitlines():
> >     proc = procs.split()
> >     if len(proc) < 6:
> >         continue
> >     if proc[4] == '--id' and proc[5].isdigit():
> >         osd_ids.append("osd.%s" % proc[5])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Single `split()` call, and is very easily grok'd.
> 
> I would prefer to have this applied, but ACK even the current code.

Thank you for the review. I have done the changes as you suggested.

**Review by pmoravec — APPROVED** [review]

**Review by TurboTurtle — APPROVED** [review]

**Review by hillpd — APPROVED** [review]
